cep-recipe-jinja.md

```diff
@@ -0,0 +1,527 @@
+- `join`: join a list with a separator (e.g. `"{{ [1, 2, 3] | join('.') }}"` will return `"1.2.3"`)
+- `min`: return the minimum value of a list (e.g. `"{{ [1, 2, 3] | min }}"` will return `1`)
+- `max`: return the maximum value of a list (e.g. `"{{ [1, 2, 3] | max }}"` will return `3`)
+- `reverse`: reverse a list (e.g. `"{{ [1, 2, 3] | reverse }}"` will return `[3, 2, 1]`)
+- `slice`: slice a list (e.g. `"{{ [1, 2, 3] | slice(1, 2) }}"` will return `[2]`)
+- `batch`: This filter works pretty much like `slice` just the other way round.
+  It returns a list of lists with the given number of items. If you provide a
+  second parameter this is used to fill up missing items.
+- `sort`: sort a list (e.g. `"{{ [3, 1, 2] | sort }}"` will return `[1, 2, 3]`)
+- `trim`: remove leading and trailing whitespace from a string (e.g. `"{{ ' foo ' | trim }}"` will return `"foo"`)
+- `unique`: remove duplicates from a list (e.g. `"{{ [1, 2, 1, 3] | unique }}"` will return `[1, 2, 3]`)
+- `split`: split a string into a list (e.g. `"{{ '1.2.3' | split('.') }}"` will return `['1', '2', '3']`). By default, splits on whitespace.
+
+<details>
+<summary>Removed filters</summary>
+The following filters are removed from the builtins:
+
+- `attr`
+- `indent` (indent with spaces, could be useful?)
+- `select`
+- `selectattr`
+- `dictsort`
+- `reject`
+- `rejectattr`
+- `round`
+- `map`
+- `title`
+- `capitalize`
+- `urlencode`
+- `escape`
+- `pprint`
+- `safe`
+- `items`
+- `float`
+- `tojson`
+</details>
+
+### Extra filters for recipes
+
+#### The `version_to_buildstring` filter
+
+- `${{ python | version_to_buildstring }}` converts a version from the variant
+  to a build string (it removes the `.` character and takes only the first two
+  elements of the version).
+
+For example the following:
+
+```yaml
+context:
+  cuda: "11.2.0"
+
+build:
+  string: ${{ hash }}_cuda${{ cuda_version | version_to_buildstring }}
+```
+
+Would evaluate to a `abc123_cuda112` (assuming the hash was `abc123`).
+
+### Various remarks
+
+#### Inline conditionals with Jinja
+
+The new recipe format allows for inline conditionals with Jinja. If they are
+falsey, and no `else` branch exists, they will render to an empty string (which
+is, for example in a list or dictionary, equivalent to a YAML `null`).
+
+When a recipe is rendered, all values that are `null` must be filtered from the
+resulting YAML.
+
+```yaml
+requirements:
+  host:
+    - ${{ "numpy" if cuda == "yes" }}
+```
+
+If `cuda` is not equal to yes, the first item of the host requirements will be
+empty (null) and thus filtered from the final list.
+
+This must also work for dictionary values. For example:
+
+```yaml
+build:
+  number: ${{ 100 if cuda == "yes" }}
+  # or an `else` branch can be used, of course
+  number: ${{ 100 if cuda == "yes" else 0 }}
+```
+
+## Error handling
+
+Build tools should be aggressive about Jinja errors:
```

Comment:
Is "should" defined somewhere in the CEP repo? If not, an implementer could interpret it as optional compared to "must".

Reply:
Ideally we determine these in a provisional CEP that others use